pydoctor/zopeinterface.py

```diff
@@ -79,49 +77,83 @@ def docsources(self) -> Iterator[model.Documentable]:
+    @param ctx: context in which C{interfaceargs} are looked up
+    """
+    for idx, arg in enumerate(interfaceargs):
+        if isinstance(arg, ast.Starred):
+            # We don't support star arguments at the moment.
+            # But these are valid, so we shouldn't warn about them either.
             continue
 
-        if isinstance(arg, str):
-            fullName = arg
+        fullName = astbuilder.node2fullname(arg, ctx)
+        if fullName is None:
+            scope.report(
+                'Interface argument %d does not look like a name' % (idx + 1),
+                section='zopeinterface')
         else:
-            fullName = scope.expandName(astor.to_source(arg).strip())
-        obj = scope.system.objForFullName(fullName)
-        if isinstance(obj, ZopeInterfaceClass):
             scope.implements_directly.append(fullName)
-            if not obj.isinterface:
-                scope.report(
-                    'probable interface %s not marked as such' % fullName,
-                    section='zopeinterface')
-                obj.isinterface = True
-                obj.kind = "Interface"
-                obj.implementedby_directly = []
-            obj.implementedby_directly.append(scope)
-        elif obj is not None:
-            scope.report(
-                'probable interface %s not detected as a class' % fullName,
+
+def _handle_implemented(
+        full_name: str,
+        implementer: Union[ZopeInterfaceClass, ZopeInterfaceModule]
+        ) -> None:
+    """This is the counterpart to addInterfaceInfoToScope(), which is called
+    during post-processing.
+    """
+    system = implementer.system
+    iface = system.objForFullName(full_name)
+    if iface is None:
+        # The interface might have been reparented, in which case there will
+        # be an alias at the original location; look for it using expandName().
+        name_parts = full_name.split('.', 1)
+        for root_obj in system.rootobjects:
+            if root_obj.name == name_parts[0]:
+                if len(name_parts) == 1:
+                    iface = root_obj
+                    break
+                iface = system.objForFullName(root_obj.expandName(name_parts[1]))
+                if iface is None:
+                    implementer.report(
+                        'Probable interface "%s" not found' % full_name,
+                        section='zopeinterface')
+                    return
+
+    if isinstance(iface, ZopeInterfaceClass):
+        if not iface.isinterface:
+            implementer.report(
+                'Probable interface "%s" not marked as such' % full_name,
                 section='zopeinterface')
+            iface.isinterface = True
+            iface.kind = "Interface"
+            iface.implementedby_directly = []
+        iface.implementedby_directly.append(implementer)
+    elif iface is not None:
+        implementer.report(
+            'Probable interface "%s" not detected as a class' % full_name,
```

Comment:
The message is a bit ambiguous (intentionally, I guess) about who is in the wrong here: it could be a problem in the code (if an actual interface isn't marked) or it could be a problem in pydoctor (if the "probable interface" isn't actually an interface).

I agree that pydoctor shouldn't aim to be a static code checker: we should only issue warnings about problems in the code that affect the quality of the generated API docs. However, that could be the case here: if the class isn't auto-converted to an interface, the generated HTML wouldn't include it in the "implements" and "implemented-by" lists.

I don't know if it's worth auto-correcting broken code. I generally prefer to have a tool issue a loud and detailed complaint and then the user can fix it. But this is code that's been around for a long time and I don't really want to change its behavior unless I'm certain it will be an improvement.

Reply:
> I don't know if it's worth auto-correcting broken code.

Don't auto-correct by default :)

I think that if any auto-correct is required, it should be implemented into separate tools / or as an option.

But for this class of errors I think is best to have a human fix them...as they are not very common and there will be no return of investment if we try to implement smart code to auto-fix it :)

Reply:
I'll remove the auto-correct code then.